instrumentation-tools/src/api_surface.py

```diff
@@ -0,0 +1,76 @@
+#  Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+#  This product includes software developed at Datadog (https://www.datadoghq.com/).
+#  Copyright 2016-Present Datadog, Inc.
+
+import re
+from math import floor
+
+DIVIDER = "#"
+TESTABLE_API_REG_EX = re.compile("^(.+)(constructor|fun)(.+)")
+ROOT_REG_EX = re.compile("^(?<!companion)(\\s*)[a-zA-z]*(class|interface|enum|object)(.*)(\\.|\\s)(.+)$")
+
+"""
+In this function we will sanitize the current API prefix as following.
+Say we have the following apiSurface file:
+
+class Foo
+  fun a
+  class Bar
+    fun b
+class AnotherFoo
+  fun c
+
+When we reach line `class AnotherFoo` the prefix will be `Foo#Bar` but the API key 
+for `fun c` should be: `AnotherFoo#fun c` therefore based on the indentation difference (2) 
+in this case we will substract 2 entities from the prefix (Foo and Bar) which will result in an empty prefix at
+this point. Going further we will attach the current entity `AnotherFoo#`
+
+"""
+def sanitize_prefix(previous_indent: int,
+                    current_indent: int,
+                    prefix: str) -> str:
```

Comment:
Maybe we could use a list instead of a String to handle this. Removing a level becomes trivial and you don't have to do a split / join operation unless you really need to

Reply:
It might be simpler indeed, I'll look into it.

Reply:
I actually think we will end up in the same situation. Using a list will solve the problem in the `if` case but will force using a `join` in the `else` case. As the `else` case will always be more often than the `if` case through amortisation the `String` solution is more optimised as it is. Maybe I am not following your idea, can you be more explicit ?

Reply:
```python
def sanitize_prefix(previous_indent: int, current_indent: int, prefix: List) -> List:
    difference = floor((current_indent - previous_indent) / INDENT_SIZE)
    if difference <= 0:
        drop_number_of_groups = 1 - difference
        return prefix[:remaining_groups]
    else:
        return prefix
```

Reply:
yes but I still have to split the prefix to get that prefix as `List` parameter. With this the split will be done all the time while with my approach the split will only be done if `difference<=0`. Also say you want to persist the `prefix` as a `List` instead of a `String` then you will have to apply the `join` for each row to add it to the `row` string. Do you see what I am saying ?